rfcs/rfc0018.md

```diff
@@ -0,0 +1,323 @@
+need to be careful not to call the same method twice.
+
+So the idea is to be able to replace this:
+
+```perl
+    my $val;
+    if (   defined $data
+        && defined $data->{deeply}
+        && defined $data->{deeply}{nested}
+        && defined $data->{deeply}{nested}[0]
+        && defined $data->{deeply}{nested}[0]{data}
+    ) {
+        $val = $data->{deeply}{nested}[0]{data}{value}
+    }
+```
+
+With this:
+
+```perl
+    my $val = $data?->{deeply}?->{nested}?->[0]?->{data}?->{value};
+```
+
+And be able to replace this:
+
+```perl
+    my $val;
+    if (defined $obj) {
+        my $tmp1 = $obj->this;
+        if (defined $tmp1) {
+            my $tmp2 = $tmp1->then;
+            if (defined $tmp2) {
+                $val = $tmp2->that;
+            }
+        }
+    }
+```
+
+With this:
+
+```perl
+    my $val = $obj?->this?->then?->that;
+```
+
+## Rationale
+
+An "optional chaining" (sometimes referred to as "null safe", "safe call",
+"safe navigation" or "optional path") operator would let developers access
+values located deep within a chain of connected references and fail
+gracefully without having to check that each of them is defined.
+
+This should result in shorter, simpler and more correct expressions whenever
+the path being accessed may be missing, such as when exploring objects and
+data structures without complete guarantees of which branches/methods are
+provided.
+
+Similar solutions have been thoroughly validated by many other popular
+programming languages like JavaScript, Kotlin, C#, Swift, TypeScript, Groovy,
+PHP, Raku, Ruby and Rust, in some cases for over 15 years now [1].
+
+## Specification
+
+The `?->` operator would behave exactly like the current dereference arrow
+`->`, interacting with the exact same things and with the same precedence,
+being completely interchangeable with it. The only difference would be that,
+whenever the lefthand side of the operator is undefined, it short-circuits
+the whole expression to an empty list `()` (which becomes `undef` in scalar
+context).
+
+One could say that:
+
+    EXPR1 ?-> EXPR2
+
+is equivalent to:
+
+    defined EXPR1 ? EXPR1->EXPR2 : ()
+
+with the important caveat that EXPR1 is only evaluated once.
+
+## Backwards Compatibility
+
+All code with `?->` currently yields a compile time syntax error, so there
+are no expected conflicts with any other syntax in Perl 5.
+
+Notable exceptions are string interpolation (where `"$foo->bar"` already
+ignores the arrow and `"$foo?->{bar}"` resolves to showing the reference
+address followed by a literal `?->{bar}`) and regular expressions (where
+`?` acts as a special character).  Optional chains should be disallowed in
+those scenarios (but see 'Open Issues' below).
+
+Static tooling may confuse the new operator with a syntax error until
+updated.
+
+Earlier versions of Perl could potentially emulate this feature with a CPAN
+module relying on something like XS::Parse::Infix (perl 5.14 onwards),
```

Comment:
Note that `XS::Parse::Infix` can't magically create new infix operators on older perls. It can only work on perls suitably patched with my `PL_infix_plugin` branch. The only other utility it has is to permit other syntax modules (e.g. `Syntax::Keyword::Match`) to use those operators.

Reply:
Thank you! I will update the RFC accordingly.